### DISCUSSION TITLE
Ipopt usage stories and papers

### DISCUSSION THREAD

**Discussion by svigerske** [coin-or/Ipopt#497] [Show and tell]
The Ipopt code has been around for 16 years, but since it is open-source we typically do not know what it is used for.

If you have used Ipopt in an interesting application or know about one, or published a paper that uses Ipopt, please share your story here, ideally with a website or citation, or send a brief e-mail to ipopt@iems.northwestern.edu. This will be very helpful to the developers to secure further funding.

If Ipopt was useful to you, this gives you a chance to show your appreciation.

**Comment by svigerske** [comment]
These success stories were collected in the Ipopt wiki before:

## My Real Online Choices

Submitted by Javier Parra-Arnau, Jagdish Prasad Achara and Claude Castelluccia, INRIA Privatics, France

[My Real Online Choices](https://myrealonlinechoices.inrialpes.fr) is a research project that aims to bring transparency and privacy protection to online advertising.
The project has developed two Web-browser plug-ins that allow users to make informed and equitable decisions regarding ad blocking and tracking.
*Myadchoices*, still a research prototype, is available for Google Chrome.
*Mytrackingchoices* has also been developed for Firefox.

*Myadchoices* is a Web technology aimed to detect behavioral advertising, that is to say, those ads which are delivered to users on the basis of their browsing histories.
To this end, the plug-in models this detection problem as an optimization problem that reflects the uncertainty in determining the information available at ad platforms and tracking companies.
Also, the plug-in estimates the uniqueness of a user from the data captured by these companies.
To conduct these two tasks, we have integrated the solvers COIN-OR Linear Programming (v.1.16.6) and COIN-OR Interior Point OPTimizer (v.3.12.3) into Myadchoices.
Both solvers ranked first in our benchmark analysis.

The integration of these two solvers into Myadchoices has been a complete success.
To the best of our knowledge, in terms of transparency, Myadchoices is the first technology to provide end-users with detailed information about behavioral advertising in real-time.


## Dynamic Modeling and Optimization platform JModelica.org

Submitted by Johan Åkesson, Lund University, Sweden

[JModelica.org](http://www.jmodelica.org) is an extensible Modelica-based open source platform for optimization, simulation and analysis of complex dynamic systems.
The main objective of the project is to create an industrially viable open source platform for optimization of Modelica models, while offering a flexible platform serving as a virtual lab for algorithm development and research.
As such, JModelica.org is intended to provide a platform for technology transfer where industrially relevant problems can inspire new research and where state of the art algorithms can be propagated form academia into industrial use.

JModelica.org at a glance:
 - Model your systems using the object-oriented and equation-based language Modelica
 - Solve your complex simulation and optimization problems using state of the art numerical algorithms
 - Automate your work in the Python scripting environment
 - Visualize your results


## Pseudospectral optimal control solver PSOPT

Submitted by Victor Becerra, University of Reading, UK.

PSOPT, an open source pseudospectral optimal control solver written in C++, uses Ipopt as its default solver.
Pseudospectral methods solve optimal control problems by approximating the time-dependent variables using orthogonal polynomials.
This allows to discretize the differential equations and continuous constraints over a grid of nodes, and to compute any integrals associated with the problem using well known quadrature formulas.
PSOPT is able to deal with single or multiphase problems with continuous time nonlinear dynamics, general endpoint constraints, nonlinear path constraints, integral constraints, interior point constraints, bounds on controls and state variables, general cost function with Lagrange and Mayer terms, free or fixed initial and final conditions, linear or nonlinear linkages between phases, and fixed or free initial and final times.

PSOPT has the following features: choice between Legendre or Chebyshev polynomial approximation, automatic scaling, automatic differentiation using the ADOL-C library, numerical differentiation by using sparse finite differences, automatic identification of the Jacobian sparsity, DAE formulation, so that differential and algebraic constraints can be implemented in the same C++ function, and an easy to use interface to GNUplot to produce graphical output.

Please visit the [PSOPT home page](https://sites.google.com/a/psopt.org/psopt/Home) for further details, documentation, and download links.


## Ipopt as a solver in APMonitor

The APMonitor software uses IPOPT as a solver for advanced process control and optimization solutions.
Optimization problems can be submitted through [APMonitor's](http://apmonitor.com) [web-based interface](http://apmonitor.ath.cx/sandbox/instructions.htm).
The solution engine is Ipopt v3.5.4 with the MUMPS linear solver.
The [modeling language documentation](http://apmonitor.ath.cx/wiki) explains the necessary equation format and provides a few example models.
Similar to the NEOS server, users from a variety of academic and industrial institutions have participated by submitting models through the online resource.


## Ipopt included in GAMS distribution

Ipopt is included as a free solver in the official [GAMS](http://www.gams.com) distribution, together with other COIN-OR solvers, thanks to Stefan Vigerske's [GAMSlinks project](https://github.com/coin-or/GAMSlinks).

## Ipopt used in SmartFolio

Submitted by Boris Gnedenko, Managing director of Modern Investment Technologies Ltd.

SmartFolio is an Excel-based asset allocation, portfolio optimization and risk management software.
It contains a wide range of advanced optimization criteria, most of which result in general NLP problems, as well as more traditional QP portfolio optimization problem.
To successfully address these issues SmartFolio uses Ipopt 3.0.1, selected for being far more stable and several orders faster than standard Excel Solver.

For more details about SmartFolio see [www.smartfolio.com](http://www.smartfolio.com).


## Solution of Mixed Integer Nonlinear Programs

Research project involving IBM and Carnegie Mellon University.

A joint project has been started to develop open source solvers for mixed integer nonlinear programs (MINLPs), and has resulted in the COIN-OR project [Bonmin](https://github.com/coin-or/Bonmin).

Ipopt is the primary NLP solver currently used in this development.
To see more details about this project, see also the [CMU-IBM Open Source MINLP Project](http://egon.cheme.cmu.edu/IBM/page.htm).


## Solution of Almost Separable Problems with Few Common Variables

Submitted by Carl Laird, Co-author of Ipopt 3.0.0.

There are a number of optimization problems that can be cast into a form that has an almost separable structure with few common variables (Optimal Design Under Uncertainty).
The linear system that is solved at each iteration of Ipopt inherits this problem structure and decomposition techniques exist for efficient solution of these large NLPs.
My primary interest in contributing to the new C++ version of Ipopt was to help build an object-oriented optimization tool that could be easily modified for the solution of large structured problems.

Currently, I have implemented a CompositeNLP that allows the building of a large almost separable problem and I am working on the decomposition technique (both serial and parallel version).


## Parameter Identification in Systems Biolgy

Submitted by James Lu, Johann Radon Institute for Computational and Applied Mathematics (RICAM)

By utilizing the adjoint solution capabilities of the SBML ODE Solver Library (SOSLib, http://www.tbi.univie.ac.at/~raim/odeSolver/), IPOPT as a local optimizer is used in combination with a scatter-search strategy to identify parameters from noisy data in highly nonlinear ODE models of gene networks.

More information can be found here, a tutorial given to the systems biology community http://www.icsb-2007.org/tutorials/index.html#PM6-INVERSE


## Computation of nonlinear scalings for reaction diffusion systems

Submitted by Martin Weiser, Zuse Institute Berlin

The discretization error in some reaction diffusion systems can be improved significantly by nonlinear scaling.
Optimal scalings can be found by solving 1D optimal control problems with highly nonlinear cost functional.
In most cases, IPOPT has been able to solve these problems with some hundred up to 400000 variables perfectly well.


## Unsupervised Learning for Exemplar-based Models

Submitted by Sebastian Nowozin, Max-Planck Institute for Biological Cybernetics, Department Empirical Inference, Tübingen

Certain unsupervised learning problems of recovering structure from finite data can be posed as finding a small set of prototypical samples.
Common examples include exemplar-based clustering, vector quantization and Gaussian mixture model density estimation.
In our approach, Ipopt is used to solve iteratively convex optimization problems for a variety of objective functions and constraint matrices.
Treating Ipopt as a black box optimization tool by means of the Ipopt Matlab interface has allowed us to quickly evaluate different objective functions easily.
(For details, see http://www.kyb.mpg.de/publications/attachments/ICML2008-Nowozin_%5B0%5D.pdf)


## EMSO Process Simulator

Submitted by Edson Cordeiro do Valle, Project Collaborator at Federal University of Rio Grande do Sul, Chemical Engineering Department, Porto Alegre, RS, Brazil

One of the free tools (but not open source) to interact with Ipopt it's the [EMSO](http://www.enq.ufrgs.br/trac/alsoc/wiki/EMSO) simulator, a project leaded by Prof. Argimiro Secchi and Phd Rafael Pelegrini Soares at Chemical Eng. Department at Federal Univ. of Rio Grande do Sul (a South Brazilian University).
EMSO is an equation-oriented dynamic and steady-state simulator and optimizer, it uses a high-level object-oriented modeling language.
This software provides a set of sparse algebra and automatic differentiation functionalities that evaluate a NLA (Non-Linear Algebraic Equations) system in a fast and efficient way.
Some NLP solvers (OPT++, Ipopt and a complex implementation) were interfaced with the software.
Ipopt is the fastest optimization solver, since it's fully compatible with EMSO sparse algebra structure.
Since EMSO uses an object oriented approach, the code is highly reusable.
Although it was designed to chemical eng. problems, it's possible to use it to model any kind of problem.
Nowadays, using the simulation and optimization "Kernel", it's possible to solve NLP, parameter estimation and data reconciliation problems.
EMSO its a free robust and efficient alternative to AMPL and GAMS. Enjoy EMSO!


## RecoVR Water Resources Manager Software for Industry

Submitted by Edson Cordeiro do Valle, Project Manager at [VRTech Tecnologias Industriais Ltda](http://www.vrtech.com.br/), Porto Alegre, RS, Brazil

The software [RECOVR](http://www.vrtech.com.br/en_us/recovr) is a tool for water and wastewater management and optimiziation in the process industries.
RECOVR approaches the problem of mass (water and wastewater) in a systematic way.
The water use circuit is modeled is a graphical environment using blocks like any other commercial simulator using its own equipments library, it includes 6 modules:

 * Historical Data Bank/Data Acquisition: It's possible to observe record and audit data such as flowrate, temperature and composition of process streams. The data can be manually entered or acquired for a data bank.

 * Reports: It's possible to monitor flowrate, temperature and composition of process streams, dividing them by plant areas and/or time in order to identify the most critical equipments considering water use and wastewater generation.

 * Data Reconciliation and Gross Error Detection: This module allow to detect fails in measurement equipments and adjust mass balances.

 * Observability and measurement instrument allocation: To estimate unmeasured streams and suggest the allocation of new measurement equipments.

 * Simulation: After the data collection and data adjust, it is possible to perform steady-state simulation to evaluate the behavior of water, wastewater and energy circuit.

 * Process Integration: After the process validation through simulation it is possible to perform the process integration, where new connections and topology will be suggested in order to minimize water use (promoting water reuse, recycling and reclaim) and wastewater discharge, respecting the constraints given by the user. This module is be based on mathematical programming techniques NLP (Non-linear programaing) and MINLP (mixed integer programming).

RECOVR uses a very efficient system of equation evaluation developed by VRTech, which performs automatic differentiation and supports sparse algebra.
Ipopt is used as main optimization routine for data reconciliation, process integration and also for simulation. See more [here](http://www.vrtech.com.br/en_us/recovr).


## Solving inverse kinematics of a humanoid robot

Submitted by Ugo Pattacini, Italian Institute of Technology - Robotics, Brain and Cognitive Sciences Dep., Genoa

In the framework of [RobotCub](http://www.robotcub.org) project, Ipopt has been profitably adopted within the [iKin](http://eris.liralab.it/iCub/dox/html/group__iKin.html) library to solve the inverse kinematics tasks for the limbs of iCub, the humanoid robotic platform developed at [IIT](http://www.iit.it): i.e. given the desired pose to be attained by the end-effector (e.g. the hand of a 7-DOF arm), Ipopt has been employed to find the corresponding joints configuration.
Indeed, among all the algorithms traditionally used for dealing with such kind of issue (DLS, CCD, ...), Ipopt turns to be more effective since it already encapsulates the notion of joints constraints and therefore it automatically ensures their compliance.
Ipopt has been easily interfaced with iKin code and the observed results demonstrate in many different working conditions that it is a powerful and realiable tool.


## ASCEND modelling system

Submitted by John Pye, [Solar Thermal Group](http://solar-thermal.anu.edu.au/), School of Engineering, Australian National University.

The ASCEND modelling system supports the IPOPT solver (with MUMPS or other dynamically-loaded linear solvers if available).
We have implemented auto-differentiation including second derivatives, and a number of test problems have been successfully solved.
Details: <http://ascendwiki.cheme.cmu.edu/IPOPT>


## ISE Chemical and Petrochemical Process Simulator

Submitted by Edson Cordeiro do Valle, Project Manager at [VRTech Tecnologias Industriais Ltda](http://www.vrtech.com.br/), Porto Alegre, RS, Brazil

[ISE](http://www.vrtech.com.br/en_us/simulador-ise/simulador-de-processo-qu-micos-e-petroqu-micos-ise.html) is an acronym for Integrated Simulation Environment.
It is a software tool for chemical and petrochemical process simulation which provides a user-friendly interface, so one can build complex simulations with just a few clicks in an intuitive way.
Built using Java, ISE uses modern technologies for thermodynamic canculations and flowsheet solving.
Tutorial videos can be found at [VRTech's page](http://www.vrtech.com.br/en_us/simulador-ise/simulador-ise-videos.html).
ISE uses Ipopt to calculate some thermodynamic properties and some equipment models for flowsheets simulation.
VRTech has also contributed to COIN-OR project with a JNI interface to communicate Java and Ipopt.


## Initial point for non linear bayesian regression models

Submitted by Victor de Buen, statistical consultant and software engineer in [Bayes Forecast](http://www.bayesforecast.com/cms/site/bayesforecast/lang/en/bf), Madrid, Spain.

We build statistical models of the business of our customers and the markets in which they are working.
It is very complex regression models with ARIMA noise, transfer functions, missing data, constraints, a priori information, hierarchical structures, etc..
User can also define arbitrary non linear filters using open source language [TOL](https://www.tol-project.org).

We are using IPOPT to find a plausible starting point to accelerate the convergence of MCMC methods that is a very hard problem, due to we treat with hundreds or thousands of variables and sparse constraints.
For now we use it only to find the optimal block main linear regression, conditioned to the rest of Gibbs blocks, but this is achieved by reducing the period of convergence of several thousands of simulations to only a few hundreds or even dozens of iterations.

We are planning to extend the API TolIpopt to calculate optimal bayesian decisions over these kind of models, and also to solve user defined problems.

**Comment by svigerske** [comment]
These papers were collected in the Ipopt wiki before:

## Ipopt Algorithm and Implementation

- J. Nocedal, A. Wächter, and R. A. Waltz, *Adaptive Barrier Strategies for Nonlinear Interior Methods*, Research Report RC 23563, IBM T. J. Watson Research Center, Yorktown, USA (March 2005; revised January 2006) [ps](http://www.research.ibm.com/people/a/andreasw/papers/adaptive.ps) [ps.gz](http://www.research.ibm.com/people/a/andreasw/papers/adaptive.ps.gz) [pdf](http://www.research.ibm.com/people/a/andreasw/papers/adaptive.pdf)

- O. Schenk, A. Wächter, M. Hagemann, *Combinatorial Approaches to the Solution of Saddle-Point Problems in Large-Scale Parallel Interior-Point Optimization*, Research Report RC 23824, IBM T. J. Watson Research Center, Yorktown, USA (December 2005) [pdf](http://www.research.ibm.com/people/a/andreasw/papers/rc23824.pdf)

- A. Wächter and L. T. Biegler, *On the Implementation of an Interior-Point Filter Line-Search Algorithm for Large-Scale Nonlinear Programming*, _Mathematical Programming_ 106(1), pp. 25-57, 2006 (As Research Report RC 23149, IBM T. J. Watson Research Center, Yorktown, USA [ps](http://www.research.ibm.com/people/a/andreasw/papers/ipopt.ps) [ps.gz](http://www.research.ibm.com/people/a/andreasw/papers/ipopt.ps.gz) [pdf](http://www.research.ibm.com/people/a/andreasw/papers/ipopt.pdf))

- A. Wächter and L. T. Biegler, *Line Search Filter Methods for Nonlinear Programming: Motivation and Global Convergence*, _SIAM Journal on Optimization_ 16(1), pp. 1-31, 2005 [ps, ps.gz, pdf] (As Research Report RC 23036, IBM T. J. Watson Research Center, Yorktown, USA: [ps](http://www.research.ibm.com/people/a/andreasw/papers/filter_global.ps), [ps.gz](http://www.research.ibm.com/people/a/andreasw/papers/filter_global.ps.gz), [pdf](http://www.research.ibm.com/people/a/andreasw/papers/filter_global.pdf))

- A. Wächter, *An Interior Point Algorithm for Large-Scale Nonlinear Optimization with Applications in Process Engineering*, _Phd Thesis_, Carnegie Mellon University, 2002 [ps](http://www.research.ibm.com/people/a/andreasw/papers/thesis.ps) [ps.gz](http://www.research.ibm.com/people/a/andreasw/papers/thesis.ps.gz) [pdf](http://www.research.ibm.com/people/a/andreasw/papers/thesis.pdf)

## General Optimization Theory

- P. Armand, J. Benoist and D. Orban, *Interpretation of Nonlinear Interior Methods as Damped Newton Methods*, _Cahier du GERAD G-2005-80_, 2005, [pdf](http://www.gerad.ca/fichiers/cahiers/G-2005-80.pdf)

- A. U. Raghunathan and L. T. Biegler, *Interior point methods for Mathematical Programs with Complementarity Constraints (MPCCs)*, _SIAM J. Optimization_ 15(3) pp. 720-750 (2005)

## Applications

### 2019

- Ferro, N.; Micheletti, S.; and Perotto, S., *POD-assisted strategies for structural topology optimization*, Computers & Mathematics with Applications, Volume 77, Issue 10, 2019, Pages 2804-2820, https://doi.org/10.1016/j.camwa.2019.01.010.

- S. Micheletti, S. Perotto, and L. Soli, *Topology optimization driven by anisotropic mesh adaptation: Towards a free-form design*, Computers & Structures, Volume 214, 2019, Pages 60-72, https://doi.org/10.1016/j.compstruc.2019.01.005

### 2018

- Winkler, A. W; Bellicoso, D. C; Hutter, M.; and Buchli, J., *Gait and Trajectory Optimization for Legged Systems through Phase-based End-Effector Parameterization*, IEEE Robotics and Automation Letters (RA-L), 2018, DOI: [10.1109/LRA.2018.2798285](http://dx.doi.org/10.1109/LRA.2018.2798285).

- S. Micheletti, S. Perotto, and M. Signorini, *Anisotropic mesh adaptation for the generalized Ambrosio–Tortorelli functional with application to brittle fracture*, Computers & Mathematics with Applications, Volume 75, Issue 6,
2018, Pages 2134-2152, https://doi.org/10.1016/j.camwa.2017.08.009

- N. Ferro, S. Micheletti, and S. Perotto, *Anisotropic mesh adaptation for crack propagation induced by a thermal shock in 2D*, Computer Methods in Applied Mechanics and Engineering, Volume 331, 2018, Pages 138-158, https://doi.org/10.1016/j.cma.2017.11.024

### 2017

- M. Santacatterina, M. Bottai, "Optimal probability weights for inference with constrained precision", Journal of the American Statistical Association, Sept 2017, DOI: [10.1080/01621459.2017.1375932](http://dx.doi.org/10.1080/01621459.2017.1375932).

- J. Parra-Arnau, J. P. Achara, C. Castelluccia, *Myadchoices: Bringing Transparency and Control to Online Advertising*, ACM Transactions on the Web, vol. 11, no. 1, March 2017, DOI: [10.1145/2996466](http://dx.doi.org/10.1145/2996466).

- Winkler, A. W; Farshidian, F.; Pardo, D.; Neunert, M.; and Buchli, J., *Fast Trajectory Optimization for Legged Robots using Vertex-based ZMP Constraints*, IEEE Robotics and Automation Letters (RA-L), 2017, DOI: [10.1109/LRA.2017.2723931](http://dx.doi.org/10.1109/LRA.2017.2723931).


### 2015

- Hallac, B., Kayvanloo, K., Hedengren, J.D., Hecker, W.C., Argyle, M.D., *An Optimized Simulation Model for Iron-Based Fischer-Tropsch Catalyst Design: Transfer Limitations as Functions of Operating and Design Conditions*, Chemical Engineering Journal, Volume 263, 1 March 2015, Pages 268–279, ISSN 1385-8947, DOI: [10.1016/j.cej.2014.10.108](http://dx.doi.org/10.1016/j.cej.2014.10.108).


### 2014

- Hedengren, J. D. and Asgharzadeh Shishavan, R., Powell, K.M., and Edgar, T.F., *Nonlinear Modeling, Estimation and Predictive Control in APMonitor*, Computers and Chemical Engineering, Volume 70, pg. 133–148, 2014, DOI: [10.1016/j.compchemeng.2014.04.013](http://dx.doi.org/10.1016/j.compchemeng.2014.04.013).

- Sun, L., Hedengren, J. D., and Beard, R. W., *Optimal Trajectory Generation using Model Predictive Control for Aerially Towed Cable Systems*, Journal of Guidance, Control, and Dynamics, Vol. 37, Issue 2, pp. 525-539, 2014, DOI: [10.2514/1.60820](http://arc.aiaa.org/doi/abs/10.2514/1.60820).


### 2013

- Jacobsen, L. T. and Hedengren, J. D., *Model Predictive Control with a Rigorous Model of a Solid Oxide Fuel Cell*, American Control Conference (ACC), Washington, DC, pp. 3747–3752, 2013.


### 2012

- P. Philipp, S. Altmannshofer: *Experimental Validation of a New Moving Horizon Estimator Approach for Networked Control Systems with Unsynchronized Clocks*, Accepted and to be presented at the 2012 American Control Conference, Montreal, Canada, 06/2012.


### 2011

- J. Olympio, *Optimal Control Problem for Low-Thrust Multiple Asteroid Tour Missions*, Journal of Guidance, Control and Dynamics, Vol. 34, No. 6, Nov-Dec. 2011.


### 2010

- U. Pattacini, F. Nori, L. Natale, G. Metta, G. Sandini,"*An Experimental Evaluation of a Novel Minimum-Jerk Cartesian Controller for Humanoid Robots*", IEEE/RSJ International Conference on Intelligent Robots and Systems, Taipei, Taiwan, October 18-22, 2010. [pdf](http://people.liralab.it/iron/Papers/conference/PattaciniEtAlIROS10.pdf).


### 2008

- J. M.M. Araujo, R.C.R. Rodrigues, J.P.B. Mota, *Determination of competitive isotherms of enantiomers by a hybrid inverse method using overloaded band profiles and the periodic state of the simulated moving-bed process*, _J. Chromatogr. A_, DOI: [10.1016/j.chroma.2008.01.019](http://dx.doi.org/10.1016/j.chroma.2008.01.019).

- R. C.R. Rodrigues, J.M.M. Araujo, J.P.B. Mota, *Two-Column Simulated Moving-Bed Process for Binary Separation*, _J. Chromatogr. A_ 1180, pp. 42-52 (2008).


### 2007

- J. P.B. Mota, I.A.A.C. Esteves, *Optimal Design and Experimental Assessment of Time-Variable Simulated Moving Bed for Gas Separation*, _Ind. Eng. Chem. Res._ 46, pp. 6978-6988 (2007).

- J. P.B. Mota, J.M.M. Araujo, R.C.R. Rodrigues, *Optimal Design of Simulated Moving-Bed Processes Under Flow-Rate Uncertainty*, _AIChE J._ 53, pp. 2630-2642 (2007).

- J. P.B. Mota, I.A.A.C. Esteves, M.F.J. Eusebio, *Synchronous and Asynchronous SMB Processes for Gas Separation*, _AIChE J._ 53, pp. 1192-1203 (2007).

- R. C.R. Rodrigues, J.M.M. Araujo, J.P.B. Mota, *Optimal design and experimental validation of synchronous, asynchronous and flow-modulated, simulated moving-bed processes using a single-column setup*, _J. Chromatogr. A_ 1162, pp. 14-23 (2007).

- R. C.R. Rodrigues, J.M.M. Araujo, M.F.J. Eusebio, J.P.B. Mota, *Experimental Assessment of Simulated Moving Bed and Varicol Processes Using a Single-Column Setup*, _J. Chromatogr. A_ 1142, pp. 69-80 (2007).

- Y. Lu, *Estimation of average backorders for an assemble-to-order system with random batch demands through extreme statistics*, _Naval Logistic Research_, 54(1), pp.33-45, 2007


### 2006

- M. Barttfeld, N. Alleborn, F. Durst, *Dynamic optimization of multiple-zone air impingement drying processes*, _Computers and Chemical Engineering_ 30, pp. 467-489, 2006

- V. M. Zavala, L.T. Biegler. *Large-Scale Parameter Estimation in Low-Density Polyethylene Tubular Reactors*,_Ind. & Eng.Chem. Res_, 45,  pp. 7867 -7881, 2006


### 2005

- P. Bonami, L. T. Biegler, A. R. Conn. G. Cornuéjols, I. E. Grossmann, C. D. Laird, J. Lee, A Lodi, F. Margot, N. Sawaya, A. Wächter, *An Algorithmic Framework For Convex Mixed Integer Nonlinear Programs*, Research Report RC 23771, IBM T. J. Watson Research Center, Yorktown, USA (November 2005) [pdf](http://www.research.ibm.com/people/a/andreasw/papers/rc23771.pdf)

- C. D. Laird, L. T. Biegler, B. van Bloemen Waanders, R. A. Bartlett, *Time Dependent Contaminant Source Determination for Municipal Water Networks Using Large Scale Optimization*, _ASCE Journal of Water Resource Management and Planning_ 131(2) p. 125 (2005)

- S. Kameswaran, G. Staus and L. T. Biegler, *Parameter Estimation of Core Flood and Reservoir Models*, _Computers and Chemical Engineering_ 29(8), pp. 1787-1800 (2005)

- Y. D. Lang, and L. T. Biegler, *Large-scale nonlinear programming with a CAPE-OPEN compliant interface*, _Chemical Engineering Research and Design (Trans. IChemE, Part A)_ 83, pp. 718-723 (2005)

- A. Wächter, C. Visweswariah and A. R. Conn, *Large-Scale Nonlinear Optimization in Circuit Tuning*, _Future Generation Computer Systems_ 21(8), pp. 1251-1262, 2005

- V. M. Zavala, A. Flores-Tlacuahuac, E. Vivaldo-Lima, *Dynamic Optimization of a Semi-Batch Reactor for Polyurethane Production*, _Chem. Eng. Sci._ 60(11),  pp. 3061-3079, 2005


### 2004

- M. B. Poku, J. D. Kelly, L. T. Biegler, *Nonlinear Programming Algorithms for Process Optimization with Many Degrees of Freedom*, _I & EC Research_ 43(21), pp. 6803-6812 (2004)

- A. Raghunathan, V. Gopal, D. Subramanian, L. T. Biegler and T. Samad, *Dynamic Optimization Strategies for 3D Conflict Resolution of Multiple Aircrafts*, _AIAA J. of Guidance, Control and Dynamics_ 27(4), pp. 586-594 (2004)

- A. U. Raghunathan, M. S. Diaz, L. T. Biegler, *An MPEC Formulation for Dynamic Optimization of Distillation Operation*, _Computers and Chemical Engineering_ 28(10), pp. 2037-2052 (2004)


### 2003

- A. U. Raghunathan, J. R. Pérez-Correa and L. T. Biegler, *Data Reconciliation and Parameter Estimation in Flux-Balance Analysis*, _Biotechnology and Bioengineering_ 84, pp. 700-709 (2003)

- T. Jockenhövel, L. T. Biegler, and A. Wächter, *Dynamic Optimization of the Tennessee Eastman Process Using the OptControlCentre*, _Computers and Chemical Engineering_ 27 (11) pp. 1513-1531, 2003


### 2002

- L. T. Biegler, A.M. Cervantes, and A. Wächter, *Advances in Simultaneous Strategies for Dynamic Process Optimization*, _Chemical Engineering Science_ 57 (4), pp. 575-593, 2002


### Preprints

- J. J. Arrieta-Camacho, L. T. Biegler, D. Subramanianm, *NMPC-Based Real-Time Coordination of Multiple Aircraft* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper1.pdf)

- J. J. Arrieta-Camacho, L. T. Biegler, *Real Time Optimal Guidance of Low-Thrust Spacecraft: An Application of Nonlinear Model Predictive Control* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper2.pdf)

- J. J. Arrieta-Camacho, L. T. Biegler, D. Subramanian, *Trajectory Control of Multiple Aircraft: An NMPC Approach* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper3.pdf)

- A. Flores-Tlacuahuac, L. T. Biegler, E. Saldivar-Guerra, *Optimal Grade Transitions in the HIPS Polymerization Process* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper5.pdf)

- A. Flores-Tlacuahuac, L. T. Biegler, *A Robust and Efficient Mixed-Integer Non-Linear Dynamic Optimization Approach for Simultaneous Design and Control* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper6.pdf)

- A. Flores Tlacuahuac, L. T. Biegler, *On the Global Dynamic Optimization of Highly Nonlinear Systems* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper7.pdf)

- Y. Kawajiri, L. T. Biegler, *Optimization Strategies for Simulated Moving Bed and PowerFeed Processes* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper13.pdf)

- Y. Kawajiri, L. T. Biegler, *Large-Scale Optimization Strategies for Simulated Moving Beds* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper14.pdf)

- C. D. Laird, L. T. Biegler, B. van Bloemen Waanders, *Real-time, Large Scale Optimization of Water Network Systems using a Subdomain Approach* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper15.pdf)

- C. D. Laird, L. T. Biegler, B. V. Bloemen Waanders, *A Mixed Integer Approach for Obtaining Unique Solutions in Source Inversion of Drinking Water Networks* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper16.pdf)

- Y. D. Lang, L. T. Biegler, *A Software Environment for Simultaneous Dynamic Optimization* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper17.pdf)

- A. U. Raghunathan, J. R. Perez-Correa, E. Agosin, L. T. Biegler, *Parameter Estimation in Metabolic Flux Balance Models for Batch Fermentation - Formulations and Solution Using Differential Variational Inequalities* [pdf](http://dynopt.cheme.cmu.edu/papers/preprint/paper19.pdf)

- V. M. Zavala, C.D. Laird, L.T. Biegler, *A Fast Computational Framework for Large-Scale Moving Horizon Estimation* [pdf](http://www.andrew.cmu.edu/user/vzavala/fast_mhe.pdf)

- V. M. Zavala, C.D. Laird, L.T. Biegler, *Fast Solvers and Rigorous Models: Can Both be Acommodated in NMPC?* [pdf](http://www.andrew.cmu.edu/user/vzavala/fast_nmpc.pdf)

- V. M. Zavala, C.D. Laird, L.T. Biegler, *Interior-Point Decomposition Approaches for Parallel Solution of Large-Scale Parameter Estimation Problems* [pdf](http://www.andrew.cmu.edu/user/vzavala/ces_ldpe.pdf)

**Comment by moorepants** [comment]
MOCO is a nice application picking up users for musculoskeletal modeling and simulation: https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008493

**Reply by moorepants** [reply]
Simbody also includes IPOPT: https://github.com/simbody/simbody

**Comment by ashutoshmahajan** [comment]
Minotaur is an open-source toolkit for developing solvers for MINLP. It uses Ipopt as the default local solver for continuous nonlinear optimization subproblems. Ipopt is used extensively in all the algorithms implemented in the toolkit, for both convex and nonconvex MINLPs.

Github page: https://github.com/minotaur-solver/minotaur
Publication: Minotaur: a mixed-integer nonlinear optimization toolkit, Mathematical Programming Computation volume 13, pages 301–338 (2021)

**Comment by sylvaticus** [comment]
We use IPOPT in a fairly large bio-economic model of the forest sector (ffsm, the French Forest Sector Model, used by decision makers at national levels in France).
Previously the model was coded in GAMS, but that was hard to keep model developing as GAMS was nice to solve the market equilibrium of forest products (traditional welfare optimization, recursive), but really inappropriate for everything else.
We hence recoded the model in C++, where everything is much more elegantly coded, and we use the IPOPT C++ interface to code the specific market optimization problem.
A version in Julia/Jump is under development.

All papers after 2014 on this page use the C++/IPOPT version ( "FFSM++"): https://ffsm-project.org/wiki/en/doc/home#published_articles

All the best,  
   Antonello Lobianco

**Comment by carolinetjes** [comment]
We used IPOPT in
Jagtenberg, C. J., & Mason, A. J. (2020). Improving fairness in ambulance planning by time sharing. European Journal of Operational Research, 280(3), 1095-1107.

We start with simulation & local search to identify a set of locally optimal ambulance configurations C, and evaluate performance for each inhabitant under each c \in C. We then use IPOPT to determine the timeshares to spend in each configuration such that the Bernoulli-Nash social welfare of the entire population is maximized.

Convexity depends on the input (specifically, the matrix obtained as a simulation result), and we provide an example in the appendix to show that this problem can be neither convex nor concave.
